File: Experimentation/JorgeVsg/signal_hound_sdk/scpi/VSG60/examples/cpp/scpi_vsg_cw_simple.cpp
#include <cassert>
#include <cstdio>
#include <Windows.h>

#include "visa.h"
#pragma comment(lib, "visa32.lib")

// This example demonstrates 
// 1. Using VISA to connect to the VSG60 software.
// 2. Configuring the software for a CW output.
// 3. Waiting for the *OPC bit to be set in the status register.

static void wait_for_opc(ViSession inst)
{
    viPrintf(inst, "*OPC\n");
    while(true) {
        int esr;
        viQueryf(inst, "*ESR?\n", "%d", &esr);
        if(esr) {
            break;
        }
        Sleep(16);
    }
}

void scpi_vsg_cw_simple()
{
    ViSession rm, inst;
    ViStatus rmStatus;

    // Get the VISA resource manager
    rmStatus = viOpenDefaultRM(&rm);
    assert(rmStatus == 0);

    // Open a session to the VSG software which must be running at this point
    ViStatus instStatus = viOpen(rm, "TCPIP::localhost::5024::SOCKET", VI_NULL, VI_NULL, &inst);
    assert(instStatus == 0);

    // For SOCKET programming, we want to tell VISA to use a terminating character 
    //   to end a read operation. In this case we want the newline character to end a 
    //   read operation. The termchar is typically set to newline by default. Here we
    //   set it for illustrative purposes.
    viSetAttribute(inst, VI_ATTR_TERMCHAR_EN, VI_TRUE);
    viSetAttribute(inst, VI_ATTR_TERMCHAR, '\n');

    viPrintf(inst, "OUTPUT ON\n");
    viPrintf(inst, "OUTPUT:MOD OFF\n");
    viPrintf(inst, "FREQ 1GHz\n");
    viPrintf(inst, "POW -30\n");

    wait_for_opc(inst);

    // Do your measurement here

    viClose(inst);

    return;
}

File: Experimentation/JorgeVsg/signal_hound_sdk/scpi/VSG60/examples/python/scpi_vsg_cw_simple.py
# This example demonstrates
# 1. Using PyVISA (https://pyvisa.readthedocs.io/) to connect to the VSG60 software
# 2. Configuring the software for a CW output
# 3. Waiting for the *OPC bit to be set in the status register

import pyvisa
from time import sleep

def wait_for_opc(inst):
    inst.write("*OPC")
    while True:
        esr = int(inst.query("*ESR?"))
        if esr:
            break
        sleep(16e-3)

def scpi_vsg_cw_simple():
    # Get the VISA resource manager
    rm = pyvisa.ResourceManager()

    # Open a session to the Spike software, Spike must be running at this point
    inst = rm.open_resource('TCPIP::localhost::5024::SOCKET')

    # For SOCKET programming, we want to tell VISA to use a terminating character
    #   to end a read and write operation.
    inst.read_termination = '\n'
    inst.write_termination = '\n'

    inst.write("OUTPUT ON")
    inst.write("OUTPUT:MOD OFF")
    inst.write("FREQ 1GHz")
    inst.write("POW -30")

    wait_for_opc(inst)

    # Do your measurement here

    # Done
    inst.close()

if __name__ == "__main__":
    scpi_vsg_cw_simple()
